ecosystem/indexer-grpc/indexer-grpc-file-store/src/processor.rs

```diff
@@ -81,145 +79,194 @@ impl Processor {
             cache_operator,
             file_store_operator,
             cache_chain_id,
-            enable_verbose_logging,
+            enable_expensive_logging,
         })
     }
 
-    // Starts the processing.
+    /// Starts the processing. The steps are
+    /// 1. Check chain id at the beginning and every step after
+    /// 2. Get the batch start version from file store metadata
+    /// 3. Start loop
+    ///   3.1 Check head from cache, decide whether we need to parallel process or just wait
+    ///   3.2 If we're ready to process, create max of 10 threads and fetch / upload data
+    ///   3.3 Update file store metadata at the end of a batch
     pub async fn run(&mut self) -> Result<()> {
         let cache_chain_id = self.cache_chain_id;
 
-        // If file store and cache chain id don't match, return an error.
-        let metadata = self
-            .file_store_operator
-            .create_default_file_store_metadata_if_absent(cache_chain_id)
-            .await
-            .context("Metadata did not match.")?;
+        let mut batch_start_version =
+            if let Some(metadata) = self.file_store_operator.get_file_store_metadata().await {
```

Comment:
when does the filestore processor upload the metadata file the first time?

Reply:
we've added a wait to the beginning of cache worker so it won't proceed unless metadata file exists. but filestore processor relies on items in the cache to upload files. is that going to cause a deadlock?

Reply:
the next PR fixes this. The simple / right way to do this is to have a set up process where the file and the cache gets seeded with the right info. B/c we want to do this in software, let's just have filestore do that. Basically cache worker should no longer update things like cache id and metadata version.